pkg/sentry/fsimpl/fuse/dev.go

```diff
@@ -46,54 +49,265 @@ type DeviceFD struct {
+
+		// We don't have the full common response header yet.
+		wantBytes := hdrLen - fd.writeCursor
+		bytesCopied, err := src.CopyIn(ctx, fd.writeBuf[fd.writeCursor:wantBytes])
+		if err != nil {
+			return 0, err
+		}
+		src = src.DropFirst(bytesCopied)
+
+		cn = int64(bytesCopied)
+		n += cn
+		fd.writeCursor += uint32(cn)
+		if fd.writeCursor == hdrLen {
+			// Have full header in the writeBuf. Use it to fetch the actual futureResponse
+			// from the device's completions map.
+			var hdr linux.FUSEHeaderOut
+			hdr.UnmarshalBytes(fd.writeBuf)
+
+			// We have the header now and so the writeBuf has served its purpose.
+			// We could reset it manually here but instead of doing that, at the
+			// end of the write, the writeCursor will be set to 0 thereby allowing
+			// the next request to overwrite whats in the buffer,
+
+			fut, ok := fd.completions[hdr.Unique]
+			if !ok {
+				// Server sent us a response for a request we never sent?
+				return 0, syserror.EINVAL
+			}
+
+			delete(fd.completions, hdr.Unique)
+
+			// Copy over the header into the future response. The rest of the payload
+			// will be copied over to the FR's data in the next iteration.
+			fut.hdr = &hdr
+			fd.writeCursorFR = fut
+
+			// Next iteration will now try read the complete request, if src has
+			// any data remaining. Otherwise we're done.
+		}
+	}
+
+	if fd.writeCursorFR != nil {
+		// See if the running task need to perform some action before returning.
+		// Since we just finished writing the future, we can be sure that
+		// getResponse generates a populated response.
+		if err := fd.noReceiverAction(ctx, fd.writeCursorFR.getResponse()); err != nil {
+			return 0, err
+		}
+
+		close(fd.writeCursorFR.ch)
+		fd.writeCursorFR = nil
+		fd.writeCursor = 0
+	}
+
+	return n, nil
+}
+
+// Readiness implements vfs.FileDescriptionImpl.Readiness.
+func (fd *DeviceFD) Readiness(mask waiter.EventMask) waiter.EventMask {
+	var ready waiter.EventMask
+	ready |= waiter.EventOut // FD is always writable
+	if !fd.queue.Empty() {
+		// Have reqs available, FD is readable.
+		ready |= waiter.EventIn
+	}
+
+	return ready & mask
 }
 
 // Seek implements vfs.FileDescriptionImpl.Seek.
 func (fd *DeviceFD) Seek(ctx context.Context, offset int64, whence int32) (int64, error) {
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
 	return 0, syserror.ENOSYS
 }
 
-// Register registers the FUSE device with vfsObj.
-func Register(vfsObj *vfs.VirtualFilesystem) error {
-	if err := vfsObj.RegisterDevice(vfs.CharDevice, linux.MISC_MAJOR, fuseDevMinor, fuseDevice{}, &vfs.RegisterDeviceOptions{
-		GroupName: "misc",
-	}); err != nil {
-		return err
+// noReceiverAction has the calling kernel.Task do some action if its known that no
+// receiver is going to be waiting on the future channel. This is to be used by:
+// FUSE_INIT.
+func (fd *DeviceFD) noReceiverAction(ctx context.Context, r *Response) error {
```

Comment:
Should this ctx be the one that was used when invoking the task? Namely, should it be the context of the task which created this request?

Reply:
Hmmmm that is surely interesting to think about. I'll leave the details for the init implementation PR, this PR should just set up some of the scaffolding. But I don't think using the requester ctx makes too much sense? What do you need from the context? The credendials? the PID? I would imagine the requester (task calling `mount(2)`) maybe *should* have the same creds as the server task reading and writing here. I'll give it more thought

Reply:
"Being the same" is an assumption. Since gVisor focuses on security heavily, it seems that removing all assumptions in our code would be preferred

Reply:
I wouldn't say we simply assume but rather figure out whether we can assert that. I took at a look at the `FUSE_INIT` PR and looks like for the processing, you only need the creds so make sure some capability is available. I know linux at some point asserts that the namespace that opens `/dev/fuse` is the same as that that initiates the `mount(2)` (https://elixir.bootlin.com/linux/latest/source/fs/fuse/inode.c#L1229). We can do something similar here, asserting the server serving this request has the same UID or GID as the mounter (there is some useful info about this in the `filesystemOptions` struct that you can use to assert this - you may need to add more - like the mounter user namespace).

Taking a step back - it would be useful to see how its done in Linux. What process credentials are used? The one that dispatches and sends it or the one thats processing the response? It makes more sense to me to infer things and assert them based on the context of the task thats writing than to use the context of a task that is no longer involved. In either case, lets move this discussion to the `FUSE_INIT` PR - This PR should just set up some of the scaffolding.